File: src/main/java/com/devahn/repository/UserRepository.java
package com.devahn.repository;

import com.devahn.domain.user.User;
import java.util.Optional;

public interface UserRepository {
    
    Optional<User> findByUsername(String username);
    
    Optional<User> findByEmail(String email);
    
    Long save(User user);
    
    boolean existsByUsername(String username);
    
    boolean existsByEmail(String email);
}

File: apps/api/src/auth/user.repository.ts
import { Injectable } from "@nestjs/common";
import { PrismaService } from "../prisma/prisma.service";

@Injectable()
export class UserRepository {
  constructor(private readonly prisma: PrismaService) {}

  async findByUsername(username: string) {
    return this.prisma.user.findUnique({ where: { username } });
  }

  async existsByUsername(username: string): Promise<boolean> {
    const count = await this.prisma.user.count({ where: { username } });
    return count > 0;
  }

  async existsByEmail(email: string): Promise<boolean> {
    const count = await this.prisma.user.count({ where: { email } });
    return count > 0;
  }

  async create(data: {
    username: string;
    password: string;
    email: string;
    role: string;
  }) {
    return this.prisma.user.create({ data });
  }
}
